### DISCUSSION TITLE
Usability issues

### DISCUSSION THREAD

**Discussion by vlcinsky** [frictionlessdata/frictionlessdata.io#738] [General]
# Table of Contents

1.  [Importing 24 CSV files to SQL database in 3 lines - great experience](#orga217900)
2.  [Import data from Excel and format the table - frustrating](#orgbb65566)
    1.  [Transform functions have very limited documentation](#org97ae380)
    2.  [Code contains many notes about incomplete implementation](#org6d81be8)
    3.  [Some transformation step functions are broken](#org5f3c3c2)
    4.  [The transform raise/reports strange errors](#org38d5756)
    5.  [Lazy evaluation makes debugging transformations difficult](#orga667fb9)
    6.  [Too short documentation about custom steps](#org589a09d)
3.  [How to improve the experience](#orge675dee)
    1.  [Document functions properly using docstring](#orgfe26990)
        1.  [Mark (in docstring) incomplete functions as experimental](#org1fbeaf9)
        2.  [Document (in docstring) all aspects of behavior](#org5146353)
    2.  [Write guides with checklists for specific coding topics](#orgf2bbb00)
    3.  [Focus on completing existing functionality](#orge6734f0)
        1.  [Turn all "implementation incomplete" notes into issues](#org3e67efa)
        2.  [Update test suite where needed](#org5b52d8d)
        3.  [Resolve the "implementation incomplete" issues](#orge31319f)

When I found Frictionless Data framework and libraries, it looked as my fulfilled dream.

<a id="orga217900"></a>

# Importing 24 CSV files to SQL database in 3 lines - great experience

Having 24 CSV files with complex relationships, I described them properly (incl. foreign keys etc) as `data/package.yaml`.

Following lines were supposed to import the files into sqlite database:
```python
from frictionless import  Package
package = Package("data/package.yaml")
package.to_sql("sqlite:///sqlite.db")
```

And it really worked incl. doing the import in proper order.

This was really astonishing experience - it does not happen often.

<a id="orgbb65566"></a>

# Import data from Excel and format the table - frustrating

My next task was to import Excel files from multiple people. There was an agreement about naming columns, but I did not expect everyone to follow the order so my plan was to have a pipeline doing steps:

1.  Import an Excel file
2.  select only fields with agreed names
3.  filter out empty records

Resulting resource (table) should be ready for further processing.

This time, I have spent few days trying to understand frictionless python package and it's usage and the experience was rather frustrating.

<a id="org97ae380"></a>

## Transform functions have very limited documentation

-   python code itself has docstring with single line simply repeating what the function name already tells
-   [Transform Steps documentation](https://framework.frictionlessdata.io/docs/guides/transform-steps) shows only sample code to run, but mostly does not tell any extra word of explanation
    -   There 40+ transformation steps and only two of them (table\_aggregate and table\_normalize) have a sentence or two of explanation
    -   The table\_normalize talks about the function "fixing dimensions". Real usage reveals, columns are really fixed properly, but empty lines are not removed so I do not understand, why there is plural for the word dimension.

<a id="org6d81be8"></a>

## Code contains many notes about incomplete implementation

Code for the functions is full of notes, which sound very dangerous, such as:
```python
    # Currently, metadata profiles are not fully finished; will require improvements
    # Some of the following step use **options - we need to review/fix it
    # We need to review table_pivot step as it's not fully implemented/tested
    # We need to review table_validate step as it's not fully implemented/tested
    # We need to review table_write step as it's not fully implemented/tested
    # We need to review how we use "target.schema.fields.clear()"
```

<a id="org5f3c3c2"></a>

## Some transformation step functions are broken

E.g. `field_filter` mess up field and value order (see issue #1155)

<a id="org38d5756"></a>

## The transform raise/reports strange errors

The errors were about things such as:

-   real number and resource reported number of rows differ
-   resource.sample property is empty

<a id="orga667fb9"></a>

## Lazy evaluation makes debugging transformations difficult

The processing is by design using lazy evaluation. This has apparent advantages, but also makes debugging rather difficult.

I am sure, core team members are aware of some working method to debug such situations effectively. It would be very handy to share it with users.

<a id="org589a09d"></a>

## Too short documentation about custom steps

[Transform Guide](https://framework.frictionlessdata.io/docs/guides/transform-guide#custom-steps) shows short custom step defined as plain function, [Step Guide](https://framework.frictionlessdata.io/docs/guides/extension/step-guide/) is using class derived from `Step`.

Apart from two different code samples there is no explanation, how to use it and what to care about.

<a id="orge675dee"></a>

# How to improve the experience

To summarize my experience with python library `frictionless`, I would say:

-   very promising
-   some features are already working excellently
-   generally under-documented
-   implementation of significant amount of functions is incomplete and this fact is not clearly declared to users

Following actions could improve the experience:

<a id="orgfe26990"></a>

## Document functions properly using docstring

For functions, which are intended for general reuse (e.g. in pipelines), take care to write proper specification of what it does in docstring.

Docstring are easily accessible to developer from IDE as helpstring for given function.

If there is only one place to put function description, it is in docstrings. There are tools to generate reference documentation from it.

<a id="org1fbeaf9"></a>

### Mark (in docstring) incomplete functions as experimental

If is a function implementation incomplete from functional point of view, declare it in docstring (very early) as "experimental" or "WIP" or similarly to inform the coder, that there might be surprises.

Missing optimization is not a reason for being marked as experimental.

<a id="org5146353"></a>

### Document (in docstring) all aspects of behavior

E.g. for `field_filter` describe, that:

-   order of names is significant as it reorders fields on output
-   when a name which is not existing in field names is used, such field will not be present in output
-   resulting table can have zero fields
-   metadata are changed in a way:
    -   schema fields are updated
    -   unique and foreign keys properties of the schema are (not?) updated if they contain fields, which are removed

Such a description would help not only a user, but also library developer as it is great specification for creating test cases.

<a id="orgf2bbb00"></a>

## Write guides with checklists for specific coding topics

Identify important coding topics and write for them instructions for developer. This shall explain principles of the task, things to take care of (e.g. using checklist listing important aspects such as "modify the data" + "update the schema incl. xxyy").

Such a text would help not only internal developers, but would help also external contributors.

Possible topics are e.g. (some already exists, but are too short at this moment):

-   transformation function or class for resource
-   transformation function or class for dataset
-   transformation function written as a function
-   transformation function written as a Step subclass
-   creating Plugin&#x2026;

<a id="orge6734f0"></a>

## Focus on completing existing functionality

Seeing very promising functionality which later turns to be broken is frustrating.

<a id="org3e67efa"></a>

### Turn all "implementation incomplete" notes into issues

Incomplete implementation (but not incomplete optimization) is really an issue.

Having such an issue in the tracker allows to:

-   describe incompleteness in detail
-   discuss it
-   focus in completing it

<a id="org5b52d8d"></a>

### Update test suite where needed

Using updated function documentation and also checklists from guides, one can make test suites more complete.

Hint: If there is some test, which would fail due to incomplete function implementation, you may mark it as expected failure until it is fixed. It also serves as very clear sign to potential user, which cases would probably not work properly.

<a id="orge31319f"></a>

### Resolve the "implementation incomplete" issues

Every completed implementation brings real value to end users.

**Comment by roll** [comment]
Thanks @vlcinsky!

We'll be working on this coming month and your suggestions will be really useful cc @shashigharti
